### PR TITLE
Add structured data and fields

### PR DESCRIPTION
**Pull request by blowmage**
This PR is still a work in progress. Many changes are coming, but the basic structure is here. Data allows for both anonymous field names, as well as duplicate field names. Please Look this over and provide feedback.

### CODE REVIEW

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb [hunk 1]
```diff
@@ -582,13 +605,13 @@ def snapshot strong: nil, timestamp: nil, staleness: nil
...
         #
-        #   users_types = db.types_for "users"
+        #   users_types = db.fields_for "users"
         #   db.insert "users", [{ id: 1, name: "Charlie", active: false },
         #                       { id: 2, name: "Harvey",  active: true }],
         #             types: users_types
         #
-        def types_for table, pairs: false
-          execute("SELECT * FROM #{table} WHERE 1 = 0").types pairs: pairs
+        def fields_for table
```
Comment (vkedia) [inline google-cloud-spanner/lib/google/cloud/spanner/client.rb]: where is this used? You can also query information schema to get this information.

Comment (blowmage) [inline google-cloud-spanner/lib/google/cloud/spanner/client.rb]: Nowhere yet. I'm trying to make it easier to insert/upsert data. What is a better wa to grab fields for a table?

**Comment by vkedia** [comment]
Better way is to query the information schema:
https://cloud.google.com/spanner/docs/information-schema

File: google-cloud-spanner/lib/google/cloud/spanner/fields.rb [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        # @private
+        def initialize types
+          @fields = []
+          if types.is_a? Array
+            types.each do |type|
+              @fields << field(type)
+            end
+          elsif types.is_a? Hash
+            types.each do |type|
+              @fields << field(type)
```
Comment (quartzmo) [inline google-cloud-spanner/lib/google/cloud/spanner/fields.rb]: Am I seeing things, or are lines 31-2 and 35-6 identical?

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb (hunk 1 above)
Comment (blowmage) [inline google-cloud-spanner/lib/google/cloud/spanner/client.rb]: This has been removed from the public API.

File: google-cloud-spanner/lib/google/cloud/spanner/fields.rb (hunk 2 above)
Comment (blowmage) [inline google-cloud-spanner/lib/google/cloud/spanner/fields.rb]: They are. Both Array and Hash use the same API. Originally those lines were not identical, but they were refactored and are the same now. They may diverge again in the future, so I felt it was best to keep them as they are for now.

**Comment by blowmage** [comment]
@quartzmo @vkedia This is ready for review.

**Pull request merged**